File: leetcode/842_Split_Array_into_Fibonacci_Sequence/842_Split_Array_into_Fibonacci_Sequence.cpp
class Solution {
public:
    vector<int> splitIntoFibonacci(string S) {
        vector<int> res = vector<int>();
        bp(S, res);
        return res;
    }
    bool bp(string str, vector<int>& res) {
        if (str.size() <= 0 && res.size() > 2) return true;
        for (int sz = 1; sz <= str.length(); sz++) {
            if (stol(str.substr(0, sz)) > 2147483647) return false;
            if (res.size() < 2 || res[res.size() - 1] + res[res.size() - 2] == stoi(str.substr(0, sz))) {
                if (str[0] == '0' && sz > 1) return false;
                res.push_back(stoi(str.substr(0, sz)));
                if (bp(str.substr(sz), res)) return true;
                res.pop_back();
            }
        }
        return false;
    }
};

File: leetcode/842_Split_Array_into_Fibonacci_Sequence/842_Split_Array_into_Fibonacci_Sequence.java
class Solution {
    public List<Integer> splitIntoFibonacci(String S) {
        List<Integer> res = new ArrayList<>();
        if (bp(S, res)) return res;
        else return new ArrayList<>();
    }
    
    private boolean bp(String str, List<Integer> res) {
        if (str.length() <= 0 && res.size() > 2) return true;
        for (int sz = 1; sz <= str.length(); sz++) {
            try {
                if (res.size() < 2 
                    || res.get(res.size() - 1) + res.get(res.size() - 2) == Integer.parseInt(str.substring(0, sz))) {
                    if (str.substring(0, sz).charAt(0) == '0' && sz > 1) return false;
                        res.add(Integer.parseInt(str.substring(0, sz)));
                    if (bp(str.substring(sz), res)) return true;
                    else res.remove(res.size() - 1);
                }
            } catch (NumberFormatException e) {
                return false;
            }
        }
        return false;
    }
}
